plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+}
+
+static gboolean
+fu_plugin_thunderbolt_wait_for_device (FuPlugin  *plugin,
+				       FuDevice  *dev,
+				       GError   **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	UpdateData    up_data = { TRUE, };
+	g_autoptr(GMainLoop) mainloop = NULL;
+	g_autoptr(GHashTable) changes = NULL;
+
+	up_data.mainloop = mainloop = g_main_loop_new (NULL, FALSE);
+	up_data.target_uuid = fu_device_get_id (dev);
+
+	/* this will limit the maximum amount of time we wait for
+	 * the device (i.e. 'dev') to re-appear. */
+	up_data.timeout_id = g_timeout_add (data->timeout,
+					    on_wait_for_device_timeout,
+					    mainloop);
+
+	/* this will capture the device added, removed, changed
+	 * signals while we are updating.  */
+	data->update_data = &up_data;
+	data->update_notify = on_wait_for_device_notify;
+
+	changes = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, g_object_unref);
+	up_data.changes = changes;
+
+	/* now we wait ... */
+	g_main_loop_run (mainloop);
+
+	/* restore original udev change handler */
+	data->update_data = NULL;
+	data->update_notify = NULL;
+
+	if (!up_data.have_device) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "timed out while waiting for device");
+	}
+
+	g_hash_table_foreach (changes, remove_leftover_devices, plugin);
+
+	return up_data.have_device;
+}
+
+/* internal interface  */
+
+void
+fu_plugin_thunderbolt_set_timeout (FuPlugin *plugin,
+				   guint     timeout_ms)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	data->timeout = timeout_ms;
+}
+
+/* virtual functions */
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+	const gchar *subsystems[] = { "thunderbolt", NULL };
+
+	data->udev = g_udev_client_new (subsystems);
+	g_signal_connect (data->udev, "uevent",
+			  G_CALLBACK (udev_uevent_cb), plugin);
+
+	data->timeout = FU_PLUGIN_THUNDERBOLT_UPDATE_TIMEOUT_MS;
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_object_unref (data->udev);
+}
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	GList *devices, *l;
+
+	devices = g_udev_client_query_by_subsystem (data->udev, "thunderbolt");
+	for (l = devices; l != NULL; l = l->next) {
+		GUdevDevice *device = l->data;
+		fu_plugin_thunderbolt_add (plugin, device);
+	}
+
+	g_list_foreach (devices, (GFunc) g_object_unref, NULL);
+	g_list_free (devices);
+
+	return TRUE;
+}
+
+
+gboolean
+fu_plugin_update_online (FuPlugin *plugin,
+			 FuDevice *dev,
+			 GBytes *blob_fw,
+			 FwupdInstallFlags flags,
+			 GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	const gchar *devpath;
+	gboolean ret;
+	guint64 status;
+	g_autoptr(GUdevDevice) udevice = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	ret = fu_plugin_thunderbolt_validate_firmware (blob_fw, &error_local);
```

Comment:
I think you could get rid of all the ret's in this function; just put the entire statement inside the `if`